src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py

```diff
@@ -267,23 +315,93 @@ def _get_partitioned_plan(self, root: TreeNode, child: TreeNode) -> SnowflakePla
         """
         score = get_complexity_score(node)
-        valid_node = (
-            self.complexity_score_lower_bound
-            < score
-            < self.complexity_score_upper_bound
-        ) and self._is_node_pipeline_breaker(node)
-        if valid_node:
+        is_valid = True
+        validity_status = CompilationStageTelemetryField.VALID_NODE_IN_PARTITION
+        if score < self.complexity_score_lower_bound:
+            is_valid = False
+            validity_status = (
+                CompilationStageTelemetryField.INVALID_DUE_TO_SCORE_BELOW_LOWER_BOUND
+            )
+
+        if score > self.complexity_score_upper_bound:
+            is_valid = False
+            validity_status = (
+                CompilationStageTelemetryField.INVALID_DUE_TO_SCORE_ABOVE_UPPER_BOUND
+            )
+
+        if is_valid and not self._is_node_pipeline_breaker(node):
+            is_valid = False
+            validity_status = (
+                CompilationStageTelemetryField.INVALID_DUE_TO_NON_PIPELINE_BREAKER
+            )
+
+        if is_valid and self._contains_external_cte_ref(node, root):
+            is_valid = False
+            validity_status = (
+                CompilationStageTelemetryField.INVALID_DUE_TO_EXTERNAL_CTE_REF
+            )
+
+        if is_valid:
             _logger.debug(
                 f"Added node of type {type(node)} with score {score} to pipeline breaker list."
             )
-        return valid_node, score
+
+        return validity_status, score
+
+    def _contains_external_cte_ref(self, node: TreeNode, root: TreeNode) -> bool:
+        """Method to check if a node contains a CTE in its subtree that is also referenced
+        by a different node that lies outside the subtree. An example situation is:
+
+                                   root
+                                /       \
+                            node1       node5
+                            /    \
+                        node2    node3
+                       /    |      |
+                   node4  SelectSnowflakePlan
+                                |
+                           SnowflakePlan
+                                |
+                           WithQueryBlock
+                                |
+                              node6
+
+        In this example, node2 contains a WithQueryBlock node that is also referenced
+        externally by node3.
+        Similarly, node3 contains a WithQueryBlock node that is also referenced externally
+        by node2.
+        However, node1 contains WithQueryBlock node that is not referenced externally.
+
+        We determine if a node contains an externally referenced CTE by comparing the
+        number of times each unique WithQueryBlock node is referenced in the subtree compared
+        to the number of times it is referenced in the root node.
+        """
+        if isinstance(node, SnowflakePlan) and isinstance(
```

Comment:
I will add explanation on why this is needed. Short version is the `SnowflakePlan` and `SelectSnowflake` plan automatically get marked as external cte ref without this special handling since their reference count for the `WithQueryBlock` they contain will always be 1. This will almost definitely be < root's count.